File: ainpc-core-plugin/src/main/java/ro/ainpc/world/mapping/MappingWandSelection.java
package ro.ainpc.world.mapping;

import java.util.Optional;

public record MappingWandSelection(MappingPoint pos1, MappingPoint pos2, MappingPoint point) {

    public MappingWandSelection {
    }

    public static MappingWandSelection empty() {
        return new MappingWandSelection(null, null, null);
    }

    public MappingWandSelection withPos1(MappingPoint value) {
        return new MappingWandSelection(value, pos2, point);
    }

    public MappingWandSelection withPos2(MappingPoint value) {
        return new MappingWandSelection(pos1, value, point);
    }

    public MappingWandSelection withPoint(MappingPoint value) {
        return new MappingWandSelection(pos1, pos2, value);
    }

    public Optional<MappingBounds> bounds() {
        if (pos1 == null || pos2 == null || !pos1.hasWorld() || !pos2.hasWorld()) {
            return Optional.empty();
        }
        if (!pos1.worldName().equalsIgnoreCase(pos2.worldName())) {
            return Optional.empty();
        }
        return Optional.of(new MappingBounds(
            pos1.worldName(),
            Math.min(pos1.x(), pos2.x()),
            Math.min(pos1.y(), pos2.y()),
            Math.min(pos1.z(), pos2.z()),
            Math.max(pos1.x(), pos2.x()),
            Math.max(pos1.y(), pos2.y()),
            Math.max(pos1.z(), pos2.z())
        ));
    }

    public boolean hasPoint() {
        return point != null && point.hasWorld();
    }

    public record MappingBounds(String worldName, int minX, int minY, int minZ, int maxX, int maxY, int maxZ) {
        public int centerX() {
            return minX + ((maxX - minX) / 2);
        }

        public int centerY() {
            return minY + ((maxY - minY) / 2);
        }

        public int centerZ() {
            return minZ + ((maxZ - minZ) / 2);
        }

        public String format() {
            return worldName + " " + minX + " " + minY + " " + minZ
                + " -> " + maxX + " " + maxY + " " + maxZ;
        }
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/world/mapping/MappingWandSelection.kt
package ro.ainpc.world.mapping

import java.util.Optional
import kotlin.math.max
import kotlin.math.min

class MappingWandSelection(
    private val pos1Value: MappingPoint?,
    private val pos2Value: MappingPoint?,
    private val pointValue: MappingPoint?
) {
    fun pos1(): MappingPoint? = pos1Value

    fun pos2(): MappingPoint? = pos2Value

    fun point(): MappingPoint? = pointValue

    fun withPos1(value: MappingPoint?): MappingWandSelection = MappingWandSelection(value, pos2Value, pointValue)

    fun withoutPos1(): MappingWandSelection = withPos1(null)

    fun withPos2(value: MappingPoint?): MappingWandSelection = MappingWandSelection(pos1Value, value, pointValue)

    fun withoutPos2(): MappingWandSelection = withPos2(null)

    fun withPoint(value: MappingPoint?): MappingWandSelection = MappingWandSelection(pos1Value, pos2Value, value)

    fun withoutPoint(): MappingWandSelection = withPoint(null)

    fun bounds(): Optional<MappingBounds> {
        val first = pos1Value
        val second = pos2Value
        if (first == null || second == null || !first.hasWorld() || !second.hasWorld()) {
            return Optional.empty()
        }
        if (!first.worldName().equals(second.worldName(), ignoreCase = true)) {
            return Optional.empty()
        }
        return Optional.of(
            MappingBounds(
                first.worldName(),
                min(first.x(), second.x()),
                min(first.y(), second.y()),
                min(first.z(), second.z()),
                max(first.x(), second.x()),
                max(first.y(), second.y()),
                max(first.z(), second.z())
            )
        )
    }

    fun hasPoint(): Boolean = pointValue != null && pointValue.hasWorld()

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is MappingWandSelection) {
            return false
        }

        return pos1Value == other.pos1Value &&
            pos2Value == other.pos2Value &&
            pointValue == other.pointValue
    }

    override fun hashCode(): Int {
        var result = pos1Value?.hashCode() ?: 0
        result = 31 * result + (pos2Value?.hashCode() ?: 0)
        result = 31 * result + (pointValue?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String =
        "MappingWandSelection[pos1=$pos1Value, pos2=$pos2Value, point=$pointValue]"

    class MappingBounds(
        worldName: String?,
        private val minXValue: Int,
        private val minYValue: Int,
        private val minZValue: Int,
        private val maxXValue: Int,
        private val maxYValue: Int,
        private val maxZValue: Int
    ) {
        private val worldNameValue = worldName?.trim().orEmpty()

        fun worldName(): String = worldNameValue

        fun minX(): Int = minXValue

        fun minY(): Int = minYValue

        fun minZ(): Int = minZValue

        fun maxX(): Int = maxXValue

        fun maxY(): Int = maxYValue

        fun maxZ(): Int = maxZValue

        fun centerX(): Int = minXValue + ((maxXValue - minXValue) / 2)

        fun centerY(): Int = minYValue + ((maxYValue - minYValue) / 2)

        fun centerZ(): Int = minZValue + ((maxZValue - minZValue) / 2)

        fun format(): String =
            "$worldNameValue $minXValue $minYValue $minZValue -> $maxXValue $maxYValue $maxZValue"

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }
            if (other !is MappingBounds) {
                return false
            }

            return worldNameValue == other.worldNameValue &&
                minXValue == other.minXValue &&
                minYValue == other.minYValue &&
                minZValue == other.minZValue &&
                maxXValue == other.maxXValue &&
                maxYValue == other.maxYValue &&
                maxZValue == other.maxZValue
        }

        override fun hashCode(): Int {
            var result = worldNameValue.hashCode()
            result = 31 * result + minXValue
            result = 31 * result + minYValue
            result = 31 * result + minZValue
            result = 31 * result + maxXValue
            result = 31 * result + maxYValue
            result = 31 * result + maxZValue
            return result
        }

        override fun toString(): String =
            "MappingBounds[worldName=$worldNameValue, minX=$minXValue, minY=$minYValue, minZ=$minZValue, " +
                "maxX=$maxXValue, maxY=$maxYValue, maxZ=$maxZValue]"
    }

    companion object {
        @JvmStatic
        fun empty(): MappingWandSelection = MappingWandSelection(null, null, null)
    }
}
